File: .skills/openclaw-skills/skills/jdchi/cli-qr-refiner/scripts/cli_qr_refiner.js
const fs = require('fs');
const { createCanvas } = require('canvas');

/**
 * QR-Renderer: Converts ASCII QR code matrices (█, ▀, ▄, space) into high-definition PNGs.
 * Useful for headless server environments where terminal output cannot be easily scanned.
 */
function renderQR(inputPath, outputPath) {
    const rawData = fs.readFileSync(inputPath, 'utf8');
    const lines = rawData.split('\n').filter(line => line.trim().length > 0);
    
    // Auto-detect matrix dimensions
    const height = lines.length;
    const width = Math.max(...lines.map(l => l.length));
    
    // Scale factor: each block character is represented as a 10x10 (or 10x20) pixel block
    const scale = 10;
    const canvas = createCanvas(width * scale, height * scale * 2); 
    const ctx = canvas.getContext('2d');

    // Background: White
    ctx.fillStyle = '#FFFFFF';
    ctx.fillRect(0, 0, canvas.width, canvas.height);
    
    // Drawing pixels: Black
    ctx.fillStyle = '#000000';
    lines.forEach((line, y) => {
        for (let x = 0; x < line.length; x++) {
            const char = line[x];
            // █: Full Black
            if (char === '█') {
                ctx.fillRect(x * scale, y * scale * 2, scale, scale * 2);
            } 
            // ▀: Upper Half Block
            else if (char === '▀') {
                ctx.fillRect(x * scale, y * scale * 2, scale, scale);
            } 
            // ▄: Lower Half Block
            else if (char === '▄') {
                ctx.fillRect(x * scale, y * scale * 2 + scale, scale, scale);
            }
        }
    });

    const buffer = canvas.toBuffer('image/png');
    fs.writeFileSync(outputPath, buffer);
    console.log(`[QR-Renderer] HD QR Code generated at: ${outputPath}`);
}

const args = process.argv.slice(2);
if (args.length < 2) {
    console.log("Usage: node qr_renderer.js <input_txt> <output_png>");
    process.exit(1);
}

renderQR(args[0], args[1]);

File: .skills/openclaw-skills/skills/jdchi/cli-qr-refiner/scripts/cli_qr_refiner.py
#!/usr/bin/env python3
"""
QR-Renderer: Converts ASCII QR code matrices (█, ▀, ▄, space) into high-definition PNGs.
Useful for headless server environments where terminal output cannot be easily scanned.

Usage:
    python3 cli_qr_refiner.py <input_txt_path> <output_png_path> [scale]

Examples:
    python3 cli_qr_refiner.py /tmp/qr_source.txt /tmp/qr_output.png
    python3 cli_qr_refiner.py /tmp/qr_source.txt /tmp/qr_output.png 15
"""

import sys
from PIL import Image, ImageDraw


def render_qr(input_path: str, output_path: str, scale: int = 10) -> None:
    """
    Render ASCII QR code (█ ▀ ▄) as high-definition PNG.

    Args:
        input_path:  Path to text file containing ASCII QR code
        output_path: Path to output PNG file
        scale:       Pixel size per character cell (default 10)
    """
    with open(input_path, 'r', encoding='utf-8') as f:
        lines = [line.rstrip('\n') for line in f if line.strip()]

    if not lines:
        print(f"[QR-Renderer] Error: empty input file '{input_path}'")
        sys.exit(1)

    height = len(lines)
    width = max(len(line) for line in lines)

    # Each char cell: width = scale, height = scale * 2
    # (half blocks ▀ ▄ are half-height, so we double the scale for proper rendering)
    cell_w = scale
    cell_h = scale * 2

    img = Image.new('RGB', (width * cell_w, height * cell_h), color='white')
    draw = ImageDraw.Draw(img)

    for y, line in enumerate(lines):
        for x, char in enumerate(line):
            x0 = x * cell_w
            y0 = y * cell_h
            x1 = x0 + cell_w
            y1 = y0 + cell_h

            if char == '█':
                # Full block: fill entire cell
                draw.rectangle([x0, y0, x1 - 1, y1 - 1], fill='black')
            elif char == '▀':
                # Upper half block: fill top half
                draw.rectangle([x0, y0, x1 - 1, y0 + cell_h // 2 - 1], fill='black')
            elif char == '▄':
                # Lower half block: fill bottom half
                draw.rectangle([x0, y0 + cell_h // 2, x1 - 1, y1 - 1], fill='black')
            # space and any other char → leave white (background)

    img.save(output_path, 'PNG')
    print(f"[QR-Renderer] HD QR Code generated at: {output_path}")


if __name__ == '__main__':
    if len(sys.argv) < 3:
        print("Usage: python3 cli_qr_refiner.py <input_txt> <output_png> [scale]")
        sys.exit(1)

    input_file = sys.argv[1]
    output_file = sys.argv[2]
    scale = int(sys.argv[3]) if len(sys.argv) > 3 else 10

    render_qr(input_file, output_file, scale)
